server/data_common/matrix_loader.py

```diff
@@ -115,14 +122,17 @@ def data_adaptor(self, location, app_config):
                     oldest = items[0]
                     oldest_cache = oldest[1][0]
                     oldest_key = oldest[0]
-                    oldest_cache.delete()
                     del self.datasets[oldest_key]
+                    delete_adaptor = oldest_cache
 
                 last_accessed = time.time()
                 loader = MatrixDataLoader(location, app_config=app_config)
                 cache_item = MatrixDataCacheItem(loader)
                 self.datasets[location] = (cache_item, last_accessed)
```

Comment:
The code exits the critical section (`with self.lock`) with the cache_item unlocked (ie, there is no read lock on it). Given this, I believe that it is possible for another thread to remove it from the cache (and call delete()), before the code acquires the read lock.

I think you need to acquire a read lock on the cache item within the critical section (but the yield must be outside the critical section).  It also seems fine to have the delete() outside the critical section.

Reply:
Agreed, fixed.